src/Raven.Server/Documents/Archive/DocumentsArchiver.cs

```diff
@@ -0,0 +1,208 @@
+
+                var logger = LoggingSource.Instance.GetLogger<DocumentsArchiver>(database.Name);
+                if (logger.IsOperationsEnabled)
+                    logger.Operations(msg, e);
+
+                return null;
+            }
+        }
+
+        protected override Task DoWork()
+        {
+            return DoArchiveWork();
+        }
+
+        private async Task DoArchiveWork()
+        {
+            while (ArchivalConfiguration?.Disabled == false)
+            {
+                await WaitOrThrowOperationCanceled(_archivePeriod);
+
+                await ArchiveDocs();
+            }
+        }
+
+        internal Task ArchiveDocs(int? batchSize = null)
+        {
+            return ArchiveDocs(batchSize ?? BatchSize);
+        }
+        
+        private async Task ArchiveDocs(int batchSize)
+        {
+            var currentTime = _database.Time.GetUtcNow();
+            
+            try
+            {
+                // DatabaseTopology topology;
+                // using (_database.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext serverContext))
+                // using (serverContext.OpenReadTransaction())
+                // {
+                //     topology = _database.ServerStore.Cluster.ReadDatabaseTopology(serverContext, _database.Name);
+                // }
+                //
+                // var isFirstInTopology = string.Equals(topology.AllNodes.FirstOrDefault(), _database.ServerStore.NodeTag, StringComparison.OrdinalIgnoreCase);
+
+                using (_database.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                {
+                    while (true)
+                    {
+                        context.Reset();
+                        context.Renew();
+
+                        using (context.OpenReadTransaction())
+                        {
+                            var options = new ArchivalStorage.ArchivedDocumentsOptions(context, currentTime, batchSize);
+
+                            var toArchive =_database.DocumentsStorage.ArchivalStorage.GetDocumentsToArchive(options, out var duration, CancellationToken);
+
+                            if (toArchive == null || toArchive.Count == 0)
+                                return;
+
+                            var command = new ArchiveDocumentsCommand(toArchive, _database, currentTime);
+                            await _database.TxMerger.Enqueue(command);
+
+                            if (Logger.IsInfoEnabled)
+                                Logger.Info($"Successfully archived {command.DeletionCount:#,#;;0} documents in {duration.ElapsedMilliseconds:#,#;;0} ms.");
+                        }
+                    }
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // this will stop processing
+                throw;
```

Comment:
Should we throw or just return?

Reply:
We should return, it's a cancellable operation.
https://github.com/ravendb/ravendb/pull/15715/commits/17c5028e4ff287e2e13046141cd3c802dd159836